src/GraphBLAS-sharp/Implementations.fs

```diff
@@ -112,31 +112,1541 @@ type CSRMatrix<'a when 'a : struct and 'a : equality>(csrTuples: CSRFormat<'a>)
+                                            localResultBuffer <- (%plus) localResultBuffer
+                                                ((%mult) matrixValuesBuffer.[k] vectorValuesBuffer.[j])
+                                resultValuesBuffer.[i] <- localResultBuffer
+                    @>
+
+                let vectorValues = vector.Values
+                let vectorIndices = vector.Indices
+                let matrixValues = this.Values
+                let matrixRows = this.Rows
+                let matrixColumns = this.Columns
+                let resultValues = Array.init matrixColumnCount (fun _ -> zero)
+                let binder kernelP =
+                    let ndRange = _1D(matrixColumnCount)
+                    kernelP
+                        ndRange
+                        resultValues
+                        matrixValues
+                        matrixRows
+                        matrixColumns
+                        vectorValues
+                        vectorIndices
+
+                do! RunCommand command binder
+                return! ToHost resultValues
+            }
+
+        let resultValues = oclContext.RunSync workflow
+
+        let resultElements =
+            (resultIndices, resultValues)
+            ||> Array.zip
+            |> Array.filter (snd >> (<>) zero)
+
+        upcast SparseVector(resultElements, matrixColumnCount)
+
+    override this.Mxv
+        (vector: Vector<'a>)
+        (vectorMask: Mask1D option)
+        (semiring: Semiring<'a>) =
+
+        if this.ColumnCount <> vector.Length then
+            invalidArg
+                "matrix"
+                (sprintf "Argument has invalid dimension. Need %i, but given %i" vector.Length this.ColumnCount)
+
+        let mask =
+            match vectorMask with
+            | Some m ->
+                if m.Length <> vector.Length then
+                    invalidArg
+                        "mask"
+                        (sprintf "Argument has invalid dimension. Need %i, but given %i" vector.Length m.Length)
+                m
+            | _ -> Mask1D(Array.empty, vector.Length, true) // Empty complemented mask is equal to none
+
+        match vector with
+        | :? SparseVector<'a> -> this.MxCOOv (downcast vector) mask semiring
+        | _ -> failwith "Not Implemented"
+
+    member private this.EWiseAddCOO
+        (matrix: COOMatrix<'a>)
+        (mask: Mask2D)
+        (semiring: Semiring<'a>) : Matrix<'a> =
+        let (BinaryOp plus) = semiring.PlusMonoid.Append
+        let zero = semiring.PlusMonoid.Zero
+
+        let leftIndices = this.Elements |> Array.map (fun (i, j, _) -> (i, j))
+        let rightIndices = matrix.Elements |> Array.map (fun (i, j, _) -> (i, j))
+
+        let commonIndices =
+            leftIndices
+            |> Array.filter (fun i ->
+                rightIndices
+                |> Array.contains i)
+
+        let leftCommonElements, leftUniqueElements =
+            this.Elements
```

Comment:
Huh. So, we suppose, that actual elements are always on the host. Thus should we load data back from GPGPU after each operation.